src/contracts/core/PaymentCoordinator.sol

```diff
@@ -110,88 +106,328 @@ contract PaymentCoordinator is
+     * @param paymentCalculationEndTimestamp The timestamp until which payments have been calculated
+     * @param activatedAt timestamp at which that the root can be claimed against
+     * @dev Only callable by the paymentUpdater
+     */
+    function submitRoot(
+        bytes32 root,
+        uint64 paymentCalculationStartTimestamp,
+        uint64 paymentCalculationEndTimestamp,
+        uint64 activatedAt
+    ) external onlyPaymentUpdater {
+        uint32 rootIndex = uint32(distributionRoots.length);
+        distributionRoots.push(
+            DistributionRoot({
+                root: root,
+                activatedAt: activatedAt,
+                paymentCalculationStartTimestamp: paymentCalculationStartTimestamp,
+                paymentCalculationEndTimestamp: paymentCalculationEndTimestamp
+            })
+        );
+        emit DistributionRootSubmitted(
+            rootIndex,
+            root,
+            paymentCalculationStartTimestamp,
+            paymentCalculationEndTimestamp,
+            activatedAt
+        );
+    }
+
+    /**
+     * @notice Create a RangePayment. Called from both `payForRange` and `payAllForRange`
+     */
     function _payForRange(RangePayment calldata rangePayment) internal returns (bytes32) {
-        require(rangePayment.strategiesAndMultlipliers.length > 0, "PaymentCoordinator.payForRange: no strategies set");
-        require(rangePayment.amount > 0, "PaymentCoordinator.payForRange: amount cannot be 0");
+        bytes32 rangePaymentHash = keccak256(abi.encode(msg.sender, paymentNonce, rangePayment));
+        require(
+            !isRangePaymentHash[msg.sender][rangePaymentHash],
+            "PaymentCoordinator._payForRange: range payment hash already submitted"
+        );
+        require(
+            rangePayment.strategiesAndMultlipliers.length > 0,
+            "PaymentCoordinator._payForRange: no strategies set"
+        );
+        require(rangePayment.amount > 0, "PaymentCoordinator._payForRange: amount cannot be 0");
         require(
             rangePayment.duration <= MAX_PAYMENT_DURATION,
-            "PaymentCoordinator.payForRange: duration exceeds MAX_PAYMENT_DURATION"
+            "PaymentCoordinator._payForRange: duration exceeds MAX_PAYMENT_DURATION"
         );
         require(
             rangePayment.duration % calculationIntervalSeconds == 0,
-            "PaymentCoordinator.payForRange: duration must be a multiple of calculationIntervalSeconds"
+            "PaymentCoordinator._payForRange: duration must be a multiple of calculationIntervalSeconds"
         );
-        // If retroactive payments enabled must at least be past LOWER_BOUND_START_RANGE,
-        // otherwise should start earliest at current block timestamp
+        if (retroactivePaymentsEnabled) {
+            require(
+                block.timestamp - LOWER_BOUND_START_RANGE <= rangePayment.startTimestamp,
```

Comment:
Can we name this to `RETROACTIVE_WINDOW` or something?

Reply:
Ok I actually think we should literally set a LOWER_BOUND on when they can send payments and a MAX_DURATION

The way things are currently coded, if AVSs dont implement payments before `m2.timestamp +  LOWER_BOUND_START_RANGE`, the cant pay operators from the beginning

Reply:
cc @ypatil12 @ChaoticWalrus

Reply:
are we OK with payments having a like, year-long duration?
should we have both of these things? 

I think we need to agree on desired behavior, rather than keep tweaking what's here haphazardly.
naming should be clear + also come from the specified behavior.
in my mind, it's something like:

- MAX_RETROACTIVE_LENGTH = max amount of time that a payment can start in the past
- PAYMENTS_GENESIS_TIME = absolute minimum start time for a payment

I can see arguments for both, but am not seeing clarity on what we need (or want).

likewise, MAX_DURATION should perhaps be MAX_FUTURE_LENGTH or something similar, if that's actually how we _want_ it to be used.

my 2c: do PAYMENTS_GENESIS_TIME, MAX_RETROACTIVE_LENGTH, and MAX_FUTURE_LENGTH -- the max duration is MAX_FUTURE_LENGTH + MAX_RETROACTIVE_LENGTH, and AVSs have to start making payments within MAX_RETROACTIVE_LENGTH of PAYMENTS_GENESIS_TIME in order to make payments that start from the genesis time.

I'm open to other ideas, I just want to clarify what is desired + why.

Reply:
Ok that actually seems pretty good

This allows for deprecating earlier versions and deleting some of our database due to MAX_RETROACTIVE_LENGTH

It allows for upgrading to newer versions due to MAX_FUTURE_LENGTH

it does place the bound on AVSs to integrate within MAX_RETROACTIVE_LENGTH of PAYMENTS_GENESIS_TIME. It also means that MAX_RETROACTIVE_LENGTH > paymentLaunch.timestamp - PAYMENTS_GENESIS_TIME. That seems ok tbh

Reply:
To summarize, we have a sliding window essentially configured by 
```
    /// @notice max amount of time that a payment can start in the past
    uint64 public immutable MAX_RETROACTIVE_LENGTH;
    /// @notice max amount of time that a payment can start in the future
    uint64 public immutable MAX_FUTURE_LENGTH;
```
But there is a absolute minimum defined by below where a startTimestamp cannot be before this.
```
    /// @notice absolute min timestamp that a payment can start at
    uint64 public immutable GENESIS_PAYMENT_TIMESTAMP;
```

And there is also a max on duration of a payment
```
    uint64 public immutable MAX_PAYMENT_DURATION;
```